ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt

```diff
@@ -571,19 +546,149 @@ private sealed class RuleExpr<out T> {
         }
 
         @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
-        private fun wrapRangeInCollection(range: LongRange): Collection<Long> {
-            if (range.first > range.last) {
+        private fun <T : Comparable<T>> wrapRangeInCollection(
+            range: ClosedRange<T>
+        ): Collection<T> {
+            if (range.start > range.endInclusive) {
                 throw IllegalArgumentException("can't use inverted range")
             }
 
-            return object : ClosedRange<Long> by range, Collection<Long> {
-                override val size: Int get() = (endInclusive - start).toInt()
-                override fun containsAll(elements: Collection<Long>) = elements.all {
+            return object : ClosedRange<T> by range, Collection<T> {
+                override fun containsAll(elements: Collection<T>) = elements.all {
                     range.contains(it)
                 }
-                override fun contains(value: Long) = range.contains(value)
-                override fun iterator() = range.iterator()
+                override fun contains(value: T) = range.contains(value)
+                override val size: Int get() { throw UnsupportedOperationException() }
+                override fun iterator(): Iterator<T> { throw UnsupportedOperationException() }
+            }
+        }
+    }
+}
+
+internal class Ip {
+    internal data class Octets(
+        private val ipString: String = "",
+        private val segments: List<UByte> = emptyList(),
+        private val backing: MutableList<UByte> = mutableListOf()
+    ) : List<UByte> by backing, Comparable<Octets> {
+        init {
+            val parsedOctets = segments.ifEmpty {
+                ipString.expandIpv6().toOctets()
```

Comment:
Why do we expand using IPv6 when there are no segments?

Reply:
We're using `ifEmpty`, so that `expandIpv6` will only execute if the `segments: List<UByte>` is empty. 

So it's either using the list of UByte or parsing the string to UByte 😊